File: iOS/dimina/DiminaKit/Container/Api/Ext/ExtBridgeAPI.swift
//
//  ExtBridgeAPI.swift
//  dimina
//

import Foundation

/// 处理 service 侧 `extBridge` 一次性调用。
///
/// service 侧调用链：
///   `extBridge({ event, module, data, success, fail, complete })`
///   → `invokeAPI(name=event, params={ module, data, success, fail, complete })`
///   → container: `methodName=event`, `param.module=模块名`
///
/// 本类不通过 `@BridgeMethod` 注册固定方法名，
/// 而是由 `DMPContainer.callBridgeMethod` 在未命中已知 API 时，
/// 检测 `param` 中存在 `"module"` 字段后调用 `handleExtBridge`。
public class ExtBridgeAPI: DMPContainerApi {

    /// 处理一次性 extBridge 调用。
    /// - Parameters:
    ///   - methodName: 即 service 侧 event 值（如 `"getUserInfo"`）
    ///   - param:      桥参数，包含 `module`、`data`、`success`、`fail`、`complete`
    ///   - env:        当前 bridge 上下文
    ///   - callback:   已由 DMPContainer 组装好的 triggerCallback 分发函数
    ///   - extModules: 宿主注册的模块表
    public static func handle(
        methodName: String,
        param: DMPBridgeParam,
        env: DMPBridgeEnv,
        callback: DMPBridgeCallback?,
        extModules: [String: DMPExtModuleHandler]
    ) {
        let map = param.getMap()
        let module = map["module"] as? String ?? ""
        let data = DMPMap(map["data"] as? [String: Any] ?? [:])

        guard let handler = extModules[module] else {
            let errMap = DMPMap(["errMsg": "extBridge:fail module \"\(module)\" not registered"])
            callback?(errMap, .fail)
            callback?(DMPMap(), .complete)
            return
        }

        let extCallback = DMPExtCallback(
            onSuccess: { result in
                result.set("errMsg", "\(methodName):ok")
                callback?(result, .success)
                callback?(DMPMap(), .complete)
            },
            onFail: { error in
                if error["errMsg"] == nil {
                    error.set("errMsg", "\(methodName):fail")
                }
                callback?(error, .fail)
                callback?(DMPMap(), .complete)
            }
        )

        _ = handler(methodName, data, extCallback)
    }
}

File: android/dimina/src/main/kotlin/com/didi/dimina/api/ext/ExtBridgeApi.kt
package com.didi.dimina.api.ext

import com.didi.dimina.api.APIResult
import com.didi.dimina.api.AsyncResult
import com.didi.dimina.api.BaseApiHandler
import com.didi.dimina.api.NoneResult
import com.didi.dimina.common.ApiUtils
import com.didi.dimina.common.LogUtils
import com.didi.dimina.ui.container.DiminaActivity
import org.json.JSONObject

/**
 * 处理 service 侧 extBridge 调用。
 *
 * service 侧调用链：
 *   extBridge({ event, module, data, success, fail, complete })
 *   → invokeAPI(name=event, params={ module, data, success, fail, complete })
 *   → container: apiName=event, params.module=模块名
 *
 * 本类不向 [com.didi.dimina.api.ApiRegistry] 注册固定 apiName，
 * 而是由 [com.didi.dimina.api.ApiRegistry] 在未命中已知 API 时，
 * 检测 params 中存在 "module" 字段后转发到此处处理。
 *
 * Author: Doslin
 */
class ExtBridgeApi(
    private val extModules: Map<String, ExtModuleHandler>,
) : BaseApiHandler() {

    private val tag = "ExtBridgeApi"

    // 不注册固定 apiName，由 ApiRegistry 动态路由
    override val apiNames: Set<String> = emptySet()

    override fun handleAction(
        activity: DiminaActivity,
        appId: String,
        apiName: String,
        params: JSONObject,
        responseCallback: (String) -> Unit,
    ): APIResult {
        val module = params.optString("module", "")
        val data = params.optJSONObject("data") ?: JSONObject()

        val handler = extModules[module]
        if (handler == null) {
            LogUtils.e(tag, "extBridge:fail module \"$module\" not registered")
            return AsyncResult(JSONObject().apply {
                put("errMsg", "extBridge:fail module \"$module\" not registered")
            })
        }

        val callback = object : ExtCallback {
            override fun onSuccess(result: JSONObject) {
                result.put("errMsg", "$apiName:ok")
                ApiUtils.invokeSuccess(params, result, responseCallback)
                ApiUtils.invokeComplete(params, responseCallback)
            }

            override fun onFail(error: JSONObject) {
                if (!error.has("errMsg")) {
                    error.put("errMsg", "$apiName:fail")
                }
                ApiUtils.invokeFail(params, error, responseCallback)
                ApiUtils.invokeComplete(params, responseCallback)
            }
        }

        return try {
            handler.handle(apiName, data, callback)
            // 回调由 handler 异步触发，返回 NoneResult 跳过 MiniApp 的默认回调逻辑
            NoneResult()
        } catch (e: Exception) {
            LogUtils.e(tag, "extBridge:fail ${e.message}")
            AsyncResult(JSONObject().apply {
                put("errMsg", "$apiName:fail ${e.message}")
            })
        }
    }
}
